CryptoPkg/Library/TlsLibMbedtls/TlsConfig.c

```diff
@@ -265,50 +272,39 @@ TlsSetCipherList (
 {
   TLS_CONNECTION            *TlsConn;
   EFI_STATUS                Status;
-  CONST TLS_CIPHER_MAPPING  **MappedCipher;
-  UINTN                     MappedCipherBytes;
-  UINTN                     MappedCipherCount;
-  UINTN                     CipherStringSize;
   UINTN                     Index;
-  CONST TLS_CIPHER_MAPPING  *Mapping;
-  CHAR8                     *CipherString;
-  CHAR8                     *CipherStringPosition;
+  UINTN                     TotalSize;
+  INT32                     *FinalCipherId;
+  UINTN                     MappedCipherCount;
 
   TlsConn = (TLS_CONNECTION *)Tls;
   if ((TlsConn == NULL) || (TlsConn->Ssl == NULL) || (CipherId == NULL)) {
     return EFI_INVALID_PARAMETER;
   }
 
-  //
-  // Allocate the MappedCipher array for recording the mappings that we find
-  // for the input IANA identifiers in CipherId.
-  //
   Status = SafeUintnMult (
-             CipherNum,
-             sizeof (*MappedCipher),
-             &MappedCipherBytes
+             CipherNum + 1,
+             sizeof (INT32),
+             &TotalSize
              );
   if (EFI_ERROR (Status)) {
     return EFI_OUT_OF_RESOURCES;
   }
 
-  MappedCipher = AllocatePool (MappedCipherBytes);
-  if (MappedCipher == NULL) {
+  FinalCipherId = AllocatePool (TotalSize);
+  if (FinalCipherId == NULL) {
     return EFI_OUT_OF_RESOURCES;
   }
 
   //
-  // Map the cipher IDs, and count the number of bytes for the full
-  // CipherString.
+  // Map the cipher IDs, and count the number of bytes for the full CipherString.
   //
   MappedCipherCount = 0;
-  CipherStringSize  = 0;
   for (Index = 0; Index < CipherNum; Index++) {
     //
     // Look up the IANA-to-Mbedtls mapping.
     //
-    Mapping = TlsGetCipherMapping (CipherId[Index]);
-    if (Mapping == NULL) {
+    if (TlsGetCipherMapping (CipherId[Index]) == FALSE) {

```

Comment:
Please refer to latest OpenSSL implementation, we should:
1. Get all mbedtls supported cipher
2. Compare input cipher
3. Config